### ISSUE TITLE
Relationship of this package with d3-scale’s ticks?

### ISSUE THREAD

**Issue by mbostock** [curran/ticks#2]
The existence of this package makes me wonder if it’s worth extracting this from d3-scale, perhaps into a new d3-ticks module:

https://github.com/d3/d3-scale/blob/master/src/ticks.js
https://github.com/d3/d3-scale/blob/master/src/tickFormat.js
https://github.com/d3/d3-scale/blob/master/src/nice.js

One reason I haven’t yet done this is that different scales merit different ticks. For example, a log scale has power-based ticks and a time scale has calendar-based ticks. Also, tick generation is closely-related to tick formatting because you often want a number format that is appropriate to the precision and meaning of the ticks (such as exponential formatting for log ticks).

But, you might want to have a tick-generation facility without interpolation and color and the other parts of d3-scale. You can pull parts of that out of d3-scale just by using Rollup… but maybe that’s inconvenient.

Anyway. Just wanted to mention it in case you had thoughts on how we can work together. Or if there is functionality here that should be provided in d3-scale.

**Comment by curran** [comment]
Hi Mike,

Thank you for your thoughts on this.

I had previously looked at your implementation, and I just could not wrap my head around how it is working (particularly the [use of sqrt](https://github.com/d3/d3-scale/blob/master/src/ticks.js#L3), and [this line](https://github.com/d3/d3-scale/blob/master/src/ticks.js#L16)). Perhaps I just need to stare at it longer ;)

I made this `ticks` package mainly to gain an understanding of the tick mark generation algorithm, and to experiment with the ES6 tooling and NPM. Probably this is replicating the behavior of your `ticks.js`, but I'm not sure if the behavior is exactly the same or not. It would be interesting to test if there are any differences regarding how close the number of ticks is to the requested approximate number of ticks (maybe this could even be visualized).

My main use case for `ticks` is as a dependency of [data-reduction](https://github.com/curran/data-reduction), which implements a filter -> aggregation computation for data reduction processing before visualization (work in progress). This is another thing that may have some overlap with things you are working on. I imagine that `data-reduction` could be framed as a D3 layout - it is pretty much a generalized multidimensional incarnation of [Histogram Layout](https://github.com/mbostock/d3/wiki/Histogram-Layout).

I have had that thought too about how different scales merit different ticks, and this relates to the main goal of the `data-reduction` package - unifying the aggregation computation behind visualizations like Histogram, Heatmap, and Calendar View.

I would be happy to work together in whatever ways we can on this stuff. It may indeed make more sense for  `data-reduction` to depend on `d3-scale` rather than this `ticks` package, in which case `ticks` could become deprecated in favor of `d3-scale`. I am evaluating this option.

**Comment by mbostock** [comment]
The code’s a little confusing because it’s optimized, but the principle is fairly simple: the quality of the generated ticks is evaluated based on log error. So, if you asked for 10 ticks, and we generated 20 ticks, that’s considered as bad as generating 5 ticks:

``` js
function logError(requested, actual) {
  return Math.abs(Math.log(requested) - Math.log(actual));
}

logError(10, 20); // 0.69314…
logError(10, 5); // 0.69314…
logError(10, 10); // 0
```

So, we want to generate a number of ticks that minimizes this log error. The initial estimate on the tick step is a lower bound. It’s just saying what power of 10 is the largest that will generate at least _count_ ticks in the specified domain. For example, if the domain is [1, 3] and the requested tick count is 10, then _span_ is 2 and the initial tick step is 0.1. This generates 20 ticks—too many—but it’s the initial estimate.

Once we have the initial estimate `step1`, we just have to compare it to three possible improvements:

``` js
var step1 = Math.pow(10, Math.floor(Math.log(span / count) / Math.LN10)), // 0.1
    step2 = step1 * 2, // 0.2
    step5 = step1 * 5, // 0.5
    step10 = step1 * 10; // 1.0
```

Now we want to decide which one of these is best, so we first estimate the number of ticks each candidate tick step will generate:

``` js
var count1 = span / step1, // 20
    count2 = span / step2, // 10
    count5 = span / step5, // 4
    count10 = span / step10; // 2
```

Then we convert to log error:

``` js
var error1 = logError(count, span / step1), // 0.693147180559945
    error2 = logError(count, span / step2), // 0
    error5 = logError(count, span / step5), // 0.9162907318741553
    error10 = logError(count, span / step10); // 1.6094379124341005
```

In this case, `step2` (0.2) is the best because it generates exactly 10 ticks. (Note that it probably generates 11 ticks because the domain happens to start on an integer multiple of the same power of ten, but ignore that…)

We can avoid calculating all the logs because it’s a monotonic transformation and we only care about specific thresholds: we know that step2 = step1 \* 2, step5 = step1 \* 5, and step10 = step1 \* 10. That’s where the Math.sqrt(2), Math.sqrt(10) and Math.sqrt(50) come in. The log error has a monotonic relationship with `span / count / step`, so once we calculate that value, we can derive which of the multiples of that step will be the best.

**Comment by curran** [comment]
Wow that is a really cool implementation. My implementation optimizes against a different error metric, the difference in the step size. Probably the log error is the better metric. I'm not sure I grok the whole thing completely, but I really appreciate your explanation.

Can I import the `ticks` and `nice` functions with an ES6 import statement? I'd like to do something like:

`import { ticks, nice } from "d3-scale";`

but these functions are not exported at the moment from the `index.js` of d3-scale. I can send a PR with this change if you would be open to exporting these functions.

**Comment by mbostock** [comment]
Thanks. I should try to do a real write-up of how it works sometime. This issue is a good start.

There’s not a public API for those functions yet, no (other than creating a linear scale and using linear.ticks or linear.nice). It’d be possible to make one, and I was trying to understand with this issue whether that would be worth doing.

**Comment by curran** [comment]
Indeed, this is a great start to a full write-up.

I would be happy to use the `ticks` and `nice` functions as-is as part of the generalized histogram computation thing I am working on. I made the PR before seeing this comment, haha maybe I'm too eager, and I apologize if it is incomplete - maybe I should have updated the documentation, or added tests.

Looking forward to see if you come up with another variation for a public API.

**Comment by mbostock** [comment]
I do like using scale ticks to create histogram bins, as in [bl.ocks.org/3048450](http://bl.ocks.org/mbostock/3048450). But in that case I also end up using the scale to do the visualization and render the axis, so it’s nice to have the scale itself as an object rather than being internal to the histogram.

**Comment by curran** [comment]
I have added a deprecation notice that points to d3-scale.

Also, I got [data-reduction aggregation using d3-scale ticks](https://github.com/curran/data-reduction/blob/master/index.js#L78) working. For this I ended up having to include `d3-color` in my devDependencies, which is not ideal but I think is fine.

I'm working on having data-reduction generate the input used for [heatmap](https://github.com/curran/data-reduction/blob/master/index.js#L78), such that the X or Y axes can be dynamically configured to be categorical, numeric, or date columns. Using d3-scales is perfect for this, because there are already designed to create different kinds of ticks for these different kinds of columns. I'm trying to decouple the data reduction from the visualization, so the same aggregations can be visualized in different ways.
